src/layer_env.rs

```diff
@@ -0,0 +1,882 @@
+                    let mut previous_value = result_env.get(&name).unwrap_or_default();
+
+                    if previous_value.len() > 0 {
+                        previous_value.push(self.delimiter_for(&name));
+                    }
+
+                    previous_value.push(&value);
+
+                    result_env.insert(&name, previous_value);
+                }
+                ModificationBehavior::Prepend => {
+                    let previous_value = result_env.get(&name).unwrap_or_default();
+
+                    let mut new_value = OsString::new();
+                    new_value.push(&value);
+
+                    if !previous_value.is_empty() {
+                        new_value.push(self.delimiter_for(&name));
+                        new_value.push(previous_value);
+                    }
+
+                    result_env.insert(&name, new_value);
+                }
+                _ => (),
+            };
+        }
+
+        result_env
+    }
+
+    fn delimiter_for(&self, key: impl Into<OsString>) -> OsString {
+        self.entries
+            .get(&(ModificationBehavior::Delimiter, key.into()))
+            .cloned()
+            .unwrap_or_default()
+    }
+
+    fn read_from_env_dir(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
+        let mut layer_env = Self::new();
+
+        for dir_entry in fs::read_dir(path.as_ref())? {
+            let path = dir_entry?.path();
+
+            // Rely on the Rust standard library for splitting stem and extension. Since paths
+            // are not necessarily UTF-8 encoded, this is not as trivial as it might look like.
+            // Think twice before changing this.
+            let file_name_stem = path.file_stem();
+            let file_name_extension = path.extension();
+
+            // The CNB spec explicitly states:
+            //
+            // > File contents MUST NOT be evaluated by a shell or otherwise modified before
+            // > inclusion in environment variable values.
+            // > https://github.com/buildpacks/spec/blob/a9f64de9c78022aa7a5091077a765f932d7afe42/buildpack.md#provided-by-the-buildpacks
+            //
+            // This should include parsing the contents with an assumed charset and later emitting
+            // the raw bytes of that encoding as it might change the actual data. Since this is not
+            // explicitly written in the spec, we read through the the reference implementation and
+            // determined that it also treats the file contents as raw bytes.
+            // See: https://github.com/buildpacks/lifecycle/blob/a7428a55c2a14d8a37e84285b95dc63192e3264e/env/env.go#L73-L106
+            use std::os::unix::ffi::OsStringExt;
+            let file_contents = OsString::from_vec(fs::read(&path)?);
+
+            if let Some(file_name_stem) = file_name_stem {
+                let modification_behavior = match file_name_extension {
+                    None => {
```

Comment:
Similar to the "unknown extension" question for the code below, I'm presuming we'll only ever be reading the files that libcnb wrote, and as `write_to_env_dir()` always uses an explicit file extension, it seems we should never be hitting the "no extension" case here either. This would at least mean we don't need to worry as much about the CNB API 0.5+ differences, right? Should we even go as far as to remove support for the no extension case (eg error/panic) given explicit seems preferable and it's effectively dead code?

Reply:
> I'm presuming we'll only ever be reading the files that libcnb wrote

We will also read files created by the CNB lifecycle when that layer was cached. The lifecycle will not always restore the exact files we wrote, as it will normalize some parts.

In general, I don't think this code should make any assumptions when reading data from disk. Buildpack authors could, even if we don't want them to, write arbitrary env files to disk. Another process could do the same. Maybe an older version of the same buildpack that didn't use libcnb wrote environment variables without an extension. I think staying close to the spec (or the reference implementation when the spec isn't sufficient) is very important here. Future versions of the spec/lifecycle might make assumptions about older version to ensure backwards compatibility. If we diverge from the spec, we might run into issues later.

Reply:
@Malax I had not realised the env dir was also restored when a layer was cached. If that's the case, then 100% agree.